File: leetcode/124/solution.py
"""
124. Binary Tree Maximum Path Sum
https://leetcode.com/problems/binary-tree-maximum-path-sum/
"""
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    answer: int

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]) -> int:
            if node is None:
                return 0

            # Get the maximum value of left node and right node
            # But if the value is less than 0, use 0.
            left = max(dfs(node.left), 0)
            right = max(dfs(node.right), 0)
            # Concatenate left, right, and the current value and let's call it path
            path = left + right + node.val
            # If path is greater than the current one, update it
            self.answer = max(self.answer, path)
            # Compare left and right nodes and return greater one, adding node.val
            return max(left, right) + node.val

        # Initialize self.asnwer
        self.answer = -1001
        dfs(root)

        return self.answer

File: leetcode/124/solution.go
// 124. Binary Tree Maximum Path Sum
// https://leetcode.com/problems/binary-tree-maximum-path-sum/description/

package lc124

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Solution struct {
	Root       *TreeNode
	MaxPathSum int
}

func Max(vars ...int) int {
	max := vars[0]
	for _, i := range vars {
		if max < i {
			max = i
		}
	}
	return max
}

func (s *Solution) dfs(node *TreeNode) int {
	if node == nil {
		return 0
	}

	// Get the sum of left and right nodes respectively
	// It the value is less than 0, use 0.
	left := Max(s.dfs(node.Left), 0)
	right := Max(s.dfs(node.Right), 0)
	// Populate curerntPathSum
	curerntPathSum := left + right + node.Val
	// Update answer
	s.MaxPathSum = Max(s.MaxPathSum, curerntPathSum)
	// Compare left and right nodes and return greater one with node.val
	return Max(left, right) + node.Val

}

func NewSolution(root *TreeNode) *Solution {
	s := Solution{Root: root, MaxPathSum: -1001}
	s.dfs(s.Root)
	return &s
}

func maxPathSum(root *TreeNode) int {
	s := NewSolution(root)
	return s.MaxPathSum

}
